ls1intum/Artemis
# Distributed Quiz Scheduling

### Checklist
- [x] Server: I documented the Java code using JavaDoc style. *And updated the existing documentation*
- [x] Server: K6 and all Java Tests work.


### Motivation and Context
Fix quiz scheduling and caching for distributed Artemis systems.

### Description
Moved from nested `ConcurrentHashMap`s to Halzelcast `IMap`s and from Javas `ExecutorService` to Hazelcast `IScheduledExecutorService`. This makes the cache distributed and the processing of new Submissions more fail-safe. The quiz submission caching/processing is overall slower though. This could be improved in a future PR.

### Steps for Testing

- If possible, start two instances of Artemis (make sure they are connected within a Hazelcast cluster)
- Test everything concerning quiz exercises, like
   - creating,
   - planning with start time,
   - starting manually,
   - setting visible,
   - resetting and deleting,
   - changing the checked answers frequently,
   - submitting manually and when time is up,
   - check you get the correct results,
   - check that the statistics are correct
- It would be best to test this with multiple people/accounts.
- Try to "destroy" or overload the quiz exercise service.
- (Review the Code)
